mmengine/infer/infer.py

```diff
@@ -0,0 +1,474 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import copy
+import itertools
+import os
+import os.path as osp
+from abc import ABCMeta, abstractmethod
+from datetime import datetime
+from typing import (Any, Callable, Dict, Iterable, List, Optional, Sequence,
+                    Tuple, Union)
+
+import numpy as np
+import torch
+import torch.nn as nn
+from rich.progress import track
+
+from mmengine.config import Config, ConfigDict
+from mmengine.config.utils import PKG2PROJECT
+from mmengine.dataset import COLLATE_FUNCTIONS, pseudo_collate
+from mmengine.device import get_device
+from mmengine.fileio import load
+from mmengine.registry import MODELS, VISUALIZERS, DefaultScope
+from mmengine.runner import load_checkpoint
+from mmengine.structures import InstanceData
+from mmengine.utils import get_installed_path
+from mmengine.visualization import Visualizer
+
+InstanceList = List[InstanceData]
+InputType = Union[str, np.ndarray, torch.Tensor]
+InputsType = Union[InputType, Sequence[InputType]]
+ImgType = Union[np.ndarray, Sequence[np.ndarray]]
+ResType = Union[Dict, List[Dict]]
+ConfigType = Union[Config, ConfigDict]
+ModelType = Tuple[Union[dict, ConfigType], str]
+
+
+class InferencerMeta(ABCMeta):
+    """Check the legality of the inferencer.
+
+    All Inferencer should not define duplicated keys for
+    ``preprocess_kwargs``,``forward_kwargs``, ``visualize_kwargs`` and
+    ``postprocess_kwargs``.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        assert isinstance(self.preprocess_kwargs, set)
+        assert isinstance(self.forward_kwargs, set)
+        assert isinstance(self.visualize_kwargs, set)
+        assert isinstance(self.postprocess_kwargs, set)
+
+        combinations = itertools.combinations(
+            [
+                self.preprocess_kwargs,
+                self.forward_kwargs,
+                self.visualize_kwargs,
+                self.postprocess_kwargs,
+            ],
+            2,
+        )
+
+        assert all((not x & y) for x, y in combinations), (
+            f'Class define error! {self.__name__} should not '
+            'define duplicated keys for `preprocess_kwargs`, '
+            '`forward_kwargs`, `visualize_kwargs` and '
+            '`postprocess_kwargs` are not allowed.')
```

Comment:
Here 6 union operations are performed. Would it be more efficient to check if the length of the union set of `xxx_kwargs` equals the sum of the length of each `xxx_kwargs` set?